### PR TITLE
jar: Upgrade netCDF from 4.3.19 to 4.6.5

### PR DESCRIPTION
**Pull request by rleigh-dundee**
- Upgrade jars; artefact name is now `netcdf4` rather than `netcdf`
- Add `cdm` and `httpservice` jars as well; these were split out of the main jar but are required
- Update ant and maven build to use the new versions

--------

Testing: verify that netcdf-using readers continue to function.  e.g. Imaris HDF, MINC, Veeco.  Should be handled by the full repo test, but manual testing of a couple might not hurt.

/cc @sbesson

### CODE REVIEW

**Comment by melissalinkert** [comment]
Shouldn't block the PR, but might be worth checking to see if this upgrade resolves https://trac.openmicroscopy.org/ome/ticket/12257

**Comment by sbesson** [comment]
@rleigh-dundee: can you merge `origin/develop` into this PR?

**Comment by rleigh-dundee** [comment]
@sbesson Done.

**Comment by sbesson** [comment]
It seems the POM changes break the OMERO build - see https://ci.openmicroscopy.org/job/OMERO-DEV-merge-build/ICE=3.5,jdk=8_LATEST,label=octopus/327. Excluding to rebuild the stack.

**Comment by rleigh-dundee** [comment]
@sbesson What needs doing to get this re-included?

**Comment by sbesson** [comment]
@rleigh-dundee: I would assume the original build error came from additional dependencies pulled from the `pom.xml`. I will have a look tomorrow.

**Comment by sbesson** [comment]
From the HEAD of `develop`

```
sbesson@ls30630:formats-gpl ((96e4051...)) $ mvn dependency:tree
[INFO] Scanning for projects...
...
[INFO] |     \- xerces:xercesImpl:jar:2.8.1:compile
[INFO] +- edu.ucar:netcdf:jar:4.3.19:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.6:compile
```

from the HEAD of this branch

````
sbesson@ls30630:formats-gpl ((324471d...)) $ mvn dependency:tree
[INFO] Scanning for projects...
...
[INFO] |     \- xerces:xercesImpl:jar:2.8.1:compile
[INFO] +- edu.ucar:netcdf4:jar:4.6.5:compile
[INFO] |  \- net.java.dev.jna:jna:jar:4.1.0:runtime
[INFO] +- edu.ucar:cdm:jar:4.6.5:compile
[INFO] |  +- edu.ucar:httpservices:jar:4.6.5:runtime
[INFO] |  |  +- org.apache.httpcomponents:httpclient:jar:4.5.1:runtime
[INFO] |  |  |  \- commons-codec:commons-codec:jar:1.9:runtime
[INFO] |  |  \- org.apache.httpcomponents:httpmime:jar:4.5.1:runtime
[INFO] |  +- org.apache.httpcomponents:httpcore:jar:4.4.4:runtime
[INFO] |  +- com.beust:jcommander:jar:1.35:runtime
[INFO] |  \- com.amazonaws:aws-java-sdk-s3:jar:1.10.11:runtime
[INFO] |     +- com.amazonaws:aws-java-sdk-kms:jar:1.10.11:runtime
[INFO] |     \- com.amazonaws:aws-java-sdk-core:jar:1.10.11:runtime
[INFO] |        \- com.fasterxml.jackson.core:jackson-databind:jar:2.5.3:runtime
[INFO] |           +- com.fasterxml.jackson.core:jackson-annotations:jar:2.5.0:runtime
[INFO] |           \- com.fasterxml.jackson.core:jackson-core:jar:2.5.3:runtime
[INFO] +- org.slf4j:slf4j-api:jar:1.7.6:compile
```

So I suspect we should try to update the exclusion list of the `pom.xml` to only retrieve the required dependencies first.

File: components/formats-gpl/pom.xml [hunk 1]
```diff
@@ -65,8 +65,13 @@
...
     <dependency>
       <groupId>edu.ucar</groupId>
-      <artifactId>netcdf</artifactId>
-      <version>4.3.19</version>
+      <artifactId>netcdf4</artifactId>
+      <version>4.6.5</version>
+    </dependency>
+    <dependency>
+      <groupId>edu.ucar</groupId>
+      <artifactId>cdm</artifactId>
```
Comment (sbesson) [inline components/formats-gpl/pom.xml]: I would remove this dependency declaration as it should be resolved as part of the `netcdf4` 4.6.5 dependencies and have the exclusions associated to the `netcdf4` artifact.

**Comment by sbesson** [comment]
As discussed earlier, removing the `exclude` label in favor of inline comments to control the exclusion/inclusion of this PR.

**Comment by hflynn** [comment]
Note this PR is causing https://ci.openmicroscopy.org/view/Docs/job/BIOFORMATS-DEV-merge-docs/323/warnings3Result/ - http://www.openmicroscopy.org/site/support/bio-formats5.2-staging/developers/java-library.html needs updating as part of this PR.

**Comment by sbesson** [comment]
Note with https://github.com/openmicroscopy/openmicroscopy/pull/4757, the changes in this PR will likely affect the OMERO build again (see https://github.com/openmicroscopy/bioformats/pull/2436#issuecomment-222065355 and https://github.com/openmicroscopy/openmicroscopy/pull/4736#issuecomment-231688412). Once the changes suggested in https://github.com/openmicroscopy/bioformats/pull/2436#discussion_r66914455 are pushed, we should be able to test this PR via `breaking` first to check it does not block the builds.

**Comment by sbesson** [comment]
@rleigh-dundee: let us know if you have time to push a commit for the dependency issue above. If not, this might miss the 5.2.0 deadline and will be rescheduled for the following development line.

**Comment by rleigh-codelibre** [comment]
@sbesson Removing the `cdm` dependency results in build failure, which is why I added it in the first place; the dependency isn't declared which is likely a bug unless there's some reason for that.  I checked again just now, and I can't get it to build with `cdm` removed.  The exclusions might be adjustable, but I won't have time to play with this further while on holiday.  If anyone wants to grab the branch and open a new PR then they are welcome to do so.

**Comment by sbesson** [comment]
@rleigh-dundee: thanks for the additional information. I was able to reproduce it locally. From the upstream `pom`, the issue is that `cdm` is declared as a `runtime` dependency of `netcdf 4.6.5`. This raises a few questions on which dependency should be upgraded/declared. Also I realized `netcdf 4.6.5` bumps the version of various `org.apache.httpcomponents` dependencies and this bump might need to be coordinated with OMERO which also declares these dependencies.

Given the late hour and the staff availability, I do not see us getting this fixed and tested in time for 5.2.0. Closing this PR and I will create a card to re-investigate as part of the next development series.

**Pull request closed without merging**